File: 704.二分查找/704.二分查找-java/704.二分查找.java
/*
 * @lc app=leetcode.cn id=704 lang=java
 * @lcpr version=30121
 *
 * [704] 二分查找
 */

// @lcpr-template-start

// @lcpr-template-end
// @lc code=start
class Solution {
    public int search(int[] nums, int target) {
        if (target < nums[0] || target > nums[nums.length - 1]) {
            return -1;
        }

        int left = 0, right = nums.length - 1;

        while (left <= right) {

            int middle = left + ((right - left) >> 1);

            if (nums[middle] > target) {
                right = middle - 1;
            } else if (nums[middle] < target) {
                left = middle + 1;
            } else if (nums[middle] == target) {
                return middle;
            }

        }

        return -1;
    }

}
//@lc code=end

// @lcpr case=start
// [1,3,5,6,8,65]\n3\n
// @lcpr case=end

File: 704.二分查找/704.二分查找-cpp/704.二分查找.cpp
// @lcpr-before-debug-begin

// @lcpr-before-debug-end

/*
 * @lc app=leetcode.cn id=704 lang=cpp
 * @lcpr version=30204
 *
 * [704] 二分查找
 */

// @lcpr-template-start
using namespace std;
#include <algorithm>
#include <array>
#include <bitset>
#include <climits>
#include <deque>
#include <functional>
#include <iostream>
#include <list>
#include <queue>
#include <stack>
#include <tuple>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>
// @lcpr-template-end
// @lc code=start
class Solution
{
public:
    int search(vector<int> &nums, int target)
    {
        int left = 0;
        int right = nums.size();
        int mid = (left + right) >> 1;
        while (left < right)
        {
            if (nums[mid] == target)
            {
                return mid;
            }
            else
            {
                if (nums[mid] > target)
                {
                    right = mid;
                    mid = (left + right) >> 1;
                }
                else if (nums[mid] < target)
                {
                    left = mid + 1;
                    mid = (left + right) >> 1;
                }
            }
        }
        return -1;
    }
};
// @lc code=end

/*
// @lcpr case=start
// [-1,0,3,5,9,12]\n9\n
// @lcpr case=end

// @lcpr case=start
// [-1,0,3,5,9,12]\n2\n
// @lcpr case=end

 */
